### ISSUE TITLE
Downgraded component not reacting to user events in a lazy loaded module using downgradeModule approach

### ISSUE THREAD

**Issue by FDIM** [angular/angular#22581]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Downgraded component that is part of a lazy loaded module is not reacting to user events when using downgradeModule approach. 

## Expected behavior
It should work normally as if it's part of the main module / not lazy loaded.

## Minimal reproduction of the problem with instructions
- go to https://fdim.github.io/angular-hybrid-lazy-load-example/example/
- click `go to lazy loaded module`  link
- click 'click to expand' and observe that component doesn't work / nothing happens

Now: 
- open developer tools in chrome, go to sources tab and locate static.js
- put a breakpoint on line 526 (`var doDowngrade = function (injector) {`) and reload the page
- on first hit app-root component is initialized and you can observe that `needsNgZone` value is true 
- on second hit, it's a component that is loaded from lazy loaded module but now `needsNgZone` is set to false.
- If you set it's value to true using console and continue - the component will work as expected. I mean clicking `click to expand` will work and reveal the text that is hidden with *ngIf

All the code related to the example can be seen in this commit: https://github.com/FDIM/angular-hybrid-lazy-load-example/commit/918d7504b73f1bda2863b94b7868bd535eacdc25

## What is the motivation / use case for changing the behavior?
To make it possible to use downgradeComponent inside lazy loaded module with a workaround described here: #17490 

In short the workaround is about setting `$$$angularInjectorController` on some parent (in e.g. route resolve) which forces downgrade component to use that value as injector and thus correctly locate the component. As I understand this is working OK when UpgradeModule is used, but not with downgradeModule. 

I propose the following fix: 
- setting `REQUIRE_INJECTOR_NEEDS_NG_ZONE` variable to `?^^$$angularInjectorNeedsNgZone` 
- using it as follows:
![image](https://user-images.githubusercontent.com/1985429/36950069-18703308-1ff1-11e8-99ae-00a77ea4094d.png)

This way it would be possible to extend previously mentioned workaround and force the variable to be true.

## Environment

Angular version: 5.2.7

Browser: any

P.S. I would have used stackblitz for the example, but the lazy loading there is not working: https://stackblitz.com/github/fdim/angular-hybrid-lazy-load-example

**Comment by FDIM** [comment]
@gkalpak since I'll be making a PR for another issue, would you accept one for this as well? provided that we agree on the solution first. This is the reason for us having a local copy of the library

**Comment by gkalpak** [comment]
I actually think this is a more general problem with multiple ng1/ng2 layers (not specific to lazy-loading).
I.e.:

```html
<ng1-comlp-a>
  <ng2-downgraded-comp-a>
    <ng1-upgraded-comp-b>
      <ng2-downgraded-comp-b> <!-- This one will not get properly change-detected. -->
```

Let me look into it.

**Comment by gkalpak** [comment]
Yes, that is indeed the case ([demo][1]).
I think we are basing our use of `needsNgZone` on wrong assumptions, which mostly hold, but break in the `ng1 > ng2 > ng1 > ng2` case. We need to refactor things a bit.

[1]: https://stackblitz.com/edit/ngupgradelite-playground-22581?file=ng1%2Fversion-stamp.component.ts

**Comment by FDIM** [comment]
Well, I tried the example and if I apply my fix (setting needsNgZone to true with debugger) it seem to work as well. 
Could it be that the value must be true when downgradeModule is used? or are you aware of something breaking if that would be the case? 
At least to me it sort of makes sense that it should if all angularjs components are running outside angular zone.

**Comment by gkalpak** [comment]
AFAICT, [this][1] is what makes the difference (and it does indeed depend on `needsNgZone`). You can also see that changing `propagateDigest` to false for the deeply nested component "fixed" CD:

```
directive(CounterComponent.selector, downgradeComponent({
  component: CounterComponent,
  propagateDigest: false,
})).
```

The AngularJS code may or may not run inside the Angular zone (e.g. an upgraded component usually runs inside the zone), so that cannot be a deciding factor.
I too believe that `needsNgZone` should be always true when `downgradeModule()` is used.

I was planning to work on a PR, but if you want to take it on, be my guest :smiley:

[1]: https://github.com/angular/angular/blob/d509bd684980f665c3ae29eaff815e756c1a1a7d/packages/upgrade/src/common/downgrade_component_adapter.ts#L155-L163

**Comment by FDIM** [comment]
Well, I'd like to close the other PR first and then if this is still pending I'll have a look :)

**Comment by FDIM** [comment]
I just tested it out, and if I pass `propagateDigest: false` it works in my example as well. 

@gkalpak if you say otherwise, I'll try to make a PR that will add another flag to the `if` you pointed out [here](https://github.com/angular/angular/blob/d509bd684980f665c3ae29eaff815e756c1a1a7d/packages/upgrade/src/common/downgrade_component_adapter.ts#L155-L163) that will check if `downgradeModule` was used tomorrow in the evening (roughly 20 hours from now)

**Comment by Naooki** [comment]
We've been facing this issue on our project, and were struggling to find the reason of this CD problem.
That's great that you've managed to find it, great job.
Would be glad to see the fix as soon as possible.
Btw, setting `propagateDigest` to `false`  for downgraded components doesn't actually solve the problem in our case... Only manual setting of `needsNgZone ` to `false` does the thing.

**Comment by FDIM** [comment]
@Naooki What if you only pass true as first argument in your case [here ](https://github.com/FDIM/angular/blob/05e82cc147480c85340fd98b1fc65035e8ed4880/packages/upgrade/src/common/downgrade_component.ts#L122)? Does it work then?

**Comment by Naooki** [comment]
@FDIM Yeah. sorry for a delayed answer.
So I tired out passing `true` as the first argument to the `setupInputs()` call, but only this didn't actually fixes the issue in my case. To make this work I've done what you said AND removed `needsNgZone` [here](https://github.com/FDIM/angular/blob/05e82cc147480c85340fd98b1fc65035e8ed4880/packages/upgrade/src/common/downgrade_component.ts#L135), as if `needsNgZone` was true.
So it appears that `needsNgZone` should be ALWAYS true in my case, basically.

**Comment by Naooki** [comment]
To be clear, my case is: 
ng2(to bootstrap module and allow the usage of ng2 services in ng1) -> ng1 ... ng1-> ng2.
Also, I've tested out a simple case with: ng1 -> ng2, and everything works just fine.
So the problem for me is more like [here](https://github.com/FDIM/angular/blob/05e82cc147480c85340fd98b1fc65035e8ed4880/packages/upgrade/src/common/downgrade_component.ts#L100). Seems like whenever I've got component structure like ... ng2 -> ng1 -> ng2 I get this `parentInjector` defined, and therefore don't set `needsNgZone` to `true`.
